OpenRA.Mods.RA2/Traits/CarrierParent.cs

```diff
@@ -0,0 +1,286 @@
+
+		[Desc("Pip color for the spawn count.")]
+		public readonly PipType PipType = PipType.Green;
+
+		[Desc("Instantly repair spawners when they return?")]
+		public readonly bool InstantRepair = true;
+
+		[GrantedConditionReference]
+		[Desc("The condition to grant to self while spawned units are loaded.",
+			"Condition can stack with multiple spawns.")]
+		public readonly string LoadedCondition = null;
+
+		[Desc("Conditions to grant when specified actors are contained inside the transport.",
+			"A dictionary of [actor id]: [condition].")]
+		public readonly Dictionary<string, string> SpawnContainConditions = new Dictionary<string, string>();
+
+		[GrantedConditionReference]
+		public IEnumerable<string> LinterSpawnContainConditions { get { return SpawnContainConditions.Values; } }
+
+		public override object Create(ActorInitializer init) { return new CarrierParent(init, this); }
+	}
+
+	public class CarrierParent : BaseSpawnerParent, IPips, ITick, INotifyAttack, INotifyBecomingIdle
+	{
+		class CarrierChildEntry : BaseSpawnerChildEntry
+		{
+			public int RearmTicks = 0;
+			public bool IsLaunched = false;
+			public new CarrierChild SpawnerChild;
+		}
+
+		readonly Dictionary<string, Stack<int>> spawnContainTokens = new Dictionary<string, Stack<int>>();
+		public readonly CarrierParentInfo CarrierParentInfo;
+
+		CarrierChildEntry[] childEntries;
+		ConditionManager conditionManager;
+		Stack<int> loadedTokens = new Stack<int>();
+
+		int respawnTicks = 0;
+
+		int launchCondition = ConditionManager.InvalidConditionToken;
+		int launchConditionTicks;
+
+		public CarrierParent(ActorInitializer init, CarrierParentInfo info)
+			: base(init, info)
+		{
+			CarrierParentInfo = info;
+		}
+
+		protected override void Created(Actor self)
+		{
+			base.Created(self);
+			conditionManager = self.Trait<ConditionManager>();
+
+			// Spawn initial load.
+			var burst = Info.InitialActorCount == -1 ? Info.Actors.Length : Info.InitialActorCount;
+			for (var i = 0; i < burst; i++)
+				Replenish(self, ChildEntries);
+		}
+
+		public override BaseSpawnerChildEntry[] CreateChildEntries(BaseSpawnerParentInfo info)
+		{
+			childEntries = new CarrierChildEntry[info.Actors.Length]; // For this class to use
+
+			for (var i = 0; i < childEntries.Length; i++)
+				childEntries[i] = new CarrierChildEntry();
+
+			return childEntries; // For the base class to use
+		}
+
+		public override void InitializeChildEntry(Actor child, BaseSpawnerChildEntry entry)
+		{
+			var carrierChildEntry = entry as CarrierChildEntry;
+			base.InitializeChildEntry(child, carrierChildEntry);
+
+			carrierChildEntry.RearmTicks = 0;
+			carrierChildEntry.IsLaunched = false;
+			carrierChildEntry.SpawnerChild = child.Trait<CarrierChild>();
+		}
+
+		void INotifyAttack.PreparingAttack(Actor self, Target target, Armament a, Barrel barrel) { }
+
+		// The rate of fire of the dummy weapon determines the launch cycle as each shot
+		// invokes Attacking()
```

Comment:
This isn't very visibly documented, but right now I don't know where to put it instead.